### ISSUE TITLE
Fix false positives for style objects in declaration-block-trailing-semicolon

### ISSUE THREAD

**Issue by lerran80** [stylelint/stylelint#3423]
Hi, I saw the request to *follow the issue template* but I could not find this issue template anywhere. So, if this is issue is to be closed for not following the template, I will be glad to reopen it properly. Just please let me know where the template is and consider creating a link to it in the readme page.

The issue is: when defining a JSX element `style` attribute in a javascript file for React/ Preact, as in the example bellow, stylelint throws a *declaration-block-trailing-semicolon* error. This is a false positive as the contents of `style` are not css, it's a javascript object.

Example:
`<span style={{ maxWidth: something }}>{something_else}</span>`

Stylelint version usded in 0.37.0

Cheers!

**Comment by hudochenkov** [comment]
> Just please let me know where the template is and consider creating a link to it in the readme page.

When you click [“New issue”](https://github.com/stylelint/stylelint/issues/new/choose) in Github interface it offers you to select template.

Please, update you're issue using a [template](https://github.com/stylelint/stylelint/issues/new?template=REPORT_A_BUG.md), so we can have better picture of your situation.

> Stylelint version usded in 0.37.0

We've never released this version. Current version is 9.3.0.

**Comment by lerran80** [comment]
Interesting, Github did not offer me a template when I opened this. Now it did when I tried again and I could not reproduce the original missing template issue. Well, here you go:

> Clearly describe the bug

When defining a JSX element style attribute in a javascript file for React/ Preact, as in the example bellow, stylelint throws a declaration-block-trailing-semicolon error. 

> Which rule, if any, is the bug related to?

declaration-block-trailing-semicolon

> What CSS is needed to reproduce the bug?

No css. That is the bug. Its's JSX actually:
```<span style={{ maxWidth: something }}>{something_else}</span>```

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "declaration-block-trailing-semicolon": "always"
  }
}
```

> Which version of stylelint are you using?

vscode-stylelint 0.37.0
At least this is what the VSCode extensions panel tells me:
![image](https://user-images.githubusercontent.com/22422447/41857605-4d1d13c0-786e-11e8-9b5d-ac8e08366e3d.png)

> How are you running stylelint: CLI, PostCSS plugin, Node API?

As a VSCode plugin

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes. It's javascript.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:
```Expected a trailing semicolon (declaration-block-trailing-semicolon)```

**Comment by jeddy3** [comment]
@lerran80 Thanks for the report and for using the template.

It looks like we need to make sure style objects are considered as declaration blocks.

**Comment by jeddy3** [comment]
This is a bit muddier than I first thought.

From [the spec](https://www.w3.org/TR/css-style-attr/#syntax):

> The value of the style attribute must match the syntax of the contents of a CSS declaration block (excluding the delimiting braces)

So, I think the parser (I guess a side-loaded `postcss-jsx` in this instance) is behaving correctly by parsing the value of the style attribute as a declaration block.

As with https://github.com/stylelint/stylelint/issues/3410#issuecomment-399658964, I think it'll have to be the users' reponsbility to only turn on the appropriate rules for each file type.

However, this does uncover a limitation of the proposed `overrides` configuration property, which applies overrides _at the file type level_ whereas a `.vue` file can contain two types of styles in one file e.g.

```html
<style>
  .component { color: red; }
</style>
<script>
  new Vue({
    el: '#demo',
    render: function (h) {
      return (
        <span style={{ maxWidth: something }}>{something_else}</span>
      )
    }
  })
</script>
```

I'm not sure what we can do about this, but we can pick this up in https://github.com/stylelint/stylelint/issues/3410.

@lerran80 As per the recommendations in https://github.com/stylelint/stylelint/issues/3410#issuecomment-399658964, you can turn the rule off or turn off linting of `.vue` files now.

**Comment by lerran80** [comment]
@jeddy3 done. Thank you!

**Comment by dantman** [comment]
> As with [#3410 (comment)](https://github.com/stylelint/stylelint/issues/3410#issuecomment-399658964), I think it'll have to be the users' responsibility to only turn on the appropriate rules for each file type.

What if a project uses both styled-components templates and JSX inline styles (for dynamic stuff)? Then there's no association between the file extension and whether the semicolon rule should be enabled?

This seriously does not seem like user error. If stylelint is applying semicolon validation to syntaxes that cannot have semicolons, then it's a stylelint bug not a configuration issue. You can say that stylelint has some deeply ingraned assumptions that make it so fixing the issue would require a hack or a rebuild of the base. And say users will need to use an overrides hack because stylelint has a known bug (which users would be valid in considering a reason to not trust stylelint). But it cannot be said that overrides is the solution or that this is a configuration issue.

For a counter example take eslint. If I have a project with JS and TypeScript mixed together, I do not need to put my typescript linting rules in an overrides or disable them in JS files. They're just properly coded so that TypeScript rules do not enforce TypeScript checks on JS.
